### ISSUE TITLE
Vote extension planning proposal

### ISSUE THREAD

**Issue by mconcat** [tendermint/tendermint#6710]
I propose using the following structure for the ABCI++ implementation of vote extensions
- [x] PR 1 #6646 - Update vote struct definitions, and ABCI proto definition boilerplates for Vote Extensions. The rationale here is that will be a several thousand line change, that would be best separated from the logical changes in review.
- [x] PR 2 https://github.com/sikkatech/tendermint/pull/35 - Make sure that the additional vote data is passed along correctly everywhere within Tendermint, and is verified in the correct spots. Namely that signed over vote data gets into the vote struct to sign over, and additional vote data gets into the header. This involves changes to the header, detailed below.
- [x] PR 3 - Implement actual ABCI++ methods, update the example applications in order to get tests working for this.
- [ ] PR 4 - Once Prepare Proposal (#6544) lands, change the APIs for prepare proposal to allow it to edit the header as required. (See below)

Below we also discuss a plan for how to update the Header structs to enable the goals of vote extension.

## New Header / Commit structs:

Note we use "header" in quotation marks below to mean the conceptual understanding of a header, not the Tendermint Header struct. The conceptual understanding of a header corresponds to all the information a lite client following the chain should receive (namely the current Tendermint `Header` struct and `LastCommit`)

### Goals

1) Allow signed over data, to be in the "header" -- namely, have things like price oracles included in the vote. We do this by having them literally adjacent, in the Commit. This is a slightly sub-optimal design, because this data would appear in lite client proofs. However, this can then be optimized out, with the usage of self-authenticating app vote data.
2) Allow self-authenticating vote data, to have its location dependent on prepare proposal. Namely, prepare proposal should have the flexibility to move the vote data between:
    - nowhere (Its just deleted, since it was an intermediate step, e.g. threshold decryption)
    - Into the block data
    - Into the "Header"
3) We need lite clients to be able to verify this app data
    
    
#### Achieving goal 1
To achieve goal (1), we make the "LastCommit" struct (the location where votes end up being in a block), include the signed_over_app_data.

`CommitSig` struct contains the vote data that is included in the Commit struct. We add a new filed `VoteExtension SignedVoteExtension` in it, where `SignedVoteExtension` is the `VoteExtension`'s signed over fields(currently only `AppDataSigned`)

#### Achieving goal 2

Prepare Proposal can already modify the block data, and it can also just delete things. So we just need to implement a way for it to add data into the header.

We can do this, by adding to the `Header` struct in Tendermint, a field called `app_data []byte`, and PrepareProposal can return a byte slice that puts whatever it wants in that field.

#### Achieving goal 3

Well this already possible in ProcessProposal's VerifyHeader method. We should switch ProcessProposal to use this method. And then VerifyHeader should just ensure that it takes in both the `Header` struct and the `Commit` struct.

co-wrote with @ValarDragon

**Comment by thanethomson** [comment]
Seems like this issue is superseded by #9396, so I'll close this one. Feel free to reopen if this is not the case.

cc @sergio-mena

**Comment by sergio-mena** [comment]
Totally agree! (this predates me, actually)
